### PR TITLE
Feature: Add slam_toolbox option for navigation2

### PR DESCRIPTION
**Pull request by stephenadhi**
Hi,

I hope the commits and coding style fit your conventions. So in this PR, I have finished your _slam_toolbox_devel_ branch. Everything is now integrated into the humble version and running smoothly using simulation.  

**Summary**

[xslocobot_descriptions]
1. lidar.urdf.xacro
- Appended missing namespace to the LaserScan topic.

[xslocobot_nav]
1. xslocobot_slam_toolbox.launch.py
- Added sync slam toolbox node in slam_toolbox.launch.py (previously only async and localization).
- Added launch argument of use_slam_toolbox: 'true' and passing slam_mode when including the nav2_bringup launch file.
- Added launch argument of map yaml file in the case of localization mode (default=' ').
2. xslocobot_nav2_bringup.launch.py 
- Galactic to Humble migration: rename the nav2_recoveries to nav2_behaviors.
- Added _tf_remappings_ to specific nodes following https://github.com/ros-planning/navigation2/tree/humble/nav2_bringup/launch.
- Added launch argument use_slam_toolbox (default: false).
- Automatic launching of nav2_nodes such as _amcl_, _lifecycle manager slam_, _lifecycle manager localization_, _map server,_ and _map saver server_, conditioned on the slam_mode and use_slam_toolbox launch arguments.

3. nav2_params.yaml
- Galactic to Humble migration: rename the nav2_recoveries to nav2_behaviors.
- Fix costmap issues (sensor out of bounds error), change default values.
- Global costmap default: height x width 20 x 20m, resolution: 0.1.
- Place robot at the centre of the global costmap (origin -10.0, -10.0).
- Local costmap default: height x width 5 x 5m, resolution: 0.05.

[xslocobot_sim]
1. xslocobot_gz_classic.rviz
- Additional RViZ config for visualizing map and costmap topics.
- Change scan topic to /locobot/scan.

**Testing**
1. Launch simulator with RViZ
`ros2 launch interbotix_xslocobot_sim xslocobot_gz_classic.launch.py robot_model:=locobot_base robot_name:=locobot use_lidar:=true use_rviz:=true use_gazebo_gui:=true
`
2. Launch slam_toolbox
`ros2 launch interbotix_xslocobot_nav xslocobot_slam_toolbox.launch.py launch_driver:=false slam_mode:=online_sync`

image

3. Move the robot around by publishing the twist messages to the topic _/locobot/diffdrive_controller/cmd_vel_unstamped_ 

image

### CODE REVIEW

File: interbotix_ros_xslocobots/interbotix_xslocobot_nav/launch/xslocobot_nav2_bringup.launch.py [hunk 1]
```diff
@@ -31,44 +31,78 @@
     https://github.com/ros-planning/navigation2/blob/2de3f92c0f476de4bda21d1fc5268657b499b258/nav2_bringup/bringup/launch/bringup_launch.py
 """
 
+from multiprocessing import Condition
```
Comment (LSinterbotix) [inline interbotix_ros_xslocobots/interbotix_xslocobot_nav/launch/xslocobot_nav2_bringup.launch.py:34]:
```suggestion
```

File: interbotix_ros_xslocobots/interbotix_xslocobot_nav/launch/xslocobot_nav2_bringup.launch.py [hunk 2]
```diff
@@ -144,18 +178,98 @@ def launch_setup(context, *args, **kwargs):
...
         set_logging_env_var,
         controller_server_node,
         planner_server_node,
-        recoveries_server_node,
+        behavior_server_node,
         bt_navigator_node,
         waypoint_follower_node,
         lifecycle_manager_navigation_node,
+        slam_toolbox_nav2_nodes,
+
```
Comment (LSinterbotix) [inline interbotix_ros_xslocobots/interbotix_xslocobot_nav/launch/xslocobot_nav2_bringup.launch.py:272]:
```suggestion
```

File: interbotix_ros_xslocobots/interbotix_xslocobot_nav/launch/xslocobot_nav2_bringup.launch.py [hunk 3]
```diff
@@ -31,44 +31,78 @@
...
     namespace_launch_arg = LaunchConfiguration('namespace')
     use_sim_time_launch_arg = LaunchConfiguration('use_sim_time')
+    log_level_launch_arg = LaunchConfiguration('log_level')
     autostart_launch_arg = LaunchConfiguration('autostart')
+    use_respawn_launch_arg = LaunchConfiguration('use_respawn')
     nav2_params_file_launch_arg = LaunchConfiguration('nav2_params_file')
     cmd_vel_topic_launch_arg = LaunchConfiguration('cmd_vel_topic')
-
+    use_slam_toolbox_launch_arg = LaunchConfiguration('use_slam_toolbox')
+    slam_toolbox_mode_launch_arg = LaunchConfiguration('slam_toolbox_mode')
```
Comment (LSinterbotix) [inline interbotix_ros_xslocobots/interbotix_xslocobot_nav/launch/xslocobot_nav2_bringup.launch.py:70]:
```suggestion
```

File: interbotix_ros_xslocobots/interbotix_xslocobot_nav/launch/xslocobot_nav2_bringup.launch.py [hunk 4]
```diff
@@ -207,5 +321,54 @@ def generate_launch_description():
...
+            default_value='',
+            description='Full path to map yaml file to load if using localization mode'
+        )
+    )
+    declared_arguments.append(
+        DeclareLaunchArgument(
+            'log_level',
+            default_value='info',
+            choices=('debug', 'info', 'warn', 'error', 'fatal'),
+            description='log'
```
Comment (LSinterbotix) [inline interbotix_ros_xslocobots/interbotix_xslocobot_nav/launch/xslocobot_nav2_bringup.launch.py:363]:
```suggestion
            description='set the logging level of the Nav2 nodes.'
```

**Review by LSinterbotix — CHANGES_REQUESTED** [review]
Tested this locally and everything works great! Just a few minor nits and style changes before we merge.

**Comment by stephenadhi** [comment]
Hi Luke, 

I applied the changes, and should be ready to merge! 

A note regarding localization mode: I found that slam_toolbox needs to take in a pose-graph map instead of a yaml file. So I suggest only using amcl when bringing up the nav2 nodes and remove localization option offered by slam_toolbox. 

I also found more issues with nav2, for example, the new syntaxes in nav2_behavior_server and failing to plan when given a goal. 

I should be able to fix those issues by next week or so, and make new PRs. Please let me know whether you would like to actively work on these issues, so we don't work on the same thing!

Thanks,
 
Stephen

**Review by LSinterbotix — APPROVED** [review]

**Comment by LSinterbotix** [comment]
@stephenadhi 

> Please let me know whether you would like to actively work on these issues, so we don't work on the same thing!

Yes, please do! Thanks again for your great contributions!

**Pull request merged**